tests/LinkTest.php

```diff
@@ -6,10 +6,76 @@ class LinkTest extends CopilotTagTest
 {
     public function expectedWrites()
     {
+        $assetMarkdown = "\n\n[#image: /photos/123ID]|||some caption|||\n";
+
         return [
             [
-                new Link("Hello world!"),
-                "Hello world!"
+                new Link(),
+                ""
+            ],
+            [
+                new Link(""),
+                ""
+            ],
+            [
+                new Link("Hello world!", "http://li.nk"),
+                "[Hello world!](http://li.nk)"
+            ],
+            [
+                new Link("Hello world!", "http://li.nk", Array("target" => "_blank")),
+                "[Hello world!](http://li.nk){: target=\"_blank\" }"
+            ],
+            [
+                new Link("Hello world!", "http://li.nk", Array("target" => "_blank", "title" => "rawr")),
+                "[Hello world!](http://li.nk){: target=\"_blank\" title=\"rawr\" }"
+            ],
+            [
+                new Link("Hello world!", null, Array("target" => "_blank", "title" => "rawr")),
+                "[Hello world!](){: target=\"_blank\" title=\"rawr\" }"
+            ],
+            [
+                new Link("Hello world!\n\n", "http://li.nk"),
+                "[Hello world!](http://li.nk)\n\n"
+            ],
+            [
+                new Link("  Hello world!\n\n", "http://li.nk"),
+                "  [Hello world!](http://li.nk)\n\n"
+            ],
+            [
+                new Link("  Hello world!\n\n", "http://li.nk"),
+                "  [Hello world!](http://li.nk)\n\n"
+            ],
+            [
+                new Link("Hello world!", "http://li.nk     "),
+                "[Hello world!](http://li.nk)"
+            ],
+            [
+                new Link($assetMarkdown, "http://li.nk"),
+                "$assetMarkdown"
+            ],
+            [
+                new Link("Hello\nworld!", "http://li.nk"),
+                "[Hello](http://li.nk)\n[world!](http://li.nk)"
+            ],
+            [
+                new Link("{$assetMarkdown}{$assetMarkdown}", "http://li.nk"),
+                "{$assetMarkdown}{$assetMarkdown}"
+            ],
+            [
+                new Link("{$assetMarkdown}some text yo{$assetMarkdown}", "http://li.nk"),
+                "{$assetMarkdown}[some text yo](http://li.nk){$assetMarkdown}"
+            ],
+            [
+                new Link("some text yo $assetMarkdown", "http://li.nk"),
+                "[some text yo](http://li.nk) $assetMarkdown"
+            ],
+            [
+                new Link("$assetMarkdown some text yo", "http://li.nk"),
+                "{$assetMarkdown} [some text yo](http://li.nk)"
+            ],
+            [
+                new Link("this is $assetMarkdown some text yo", "http://li.nk"),
+                "[this is](http://li.nk) {$assetMarkdown} [some text yo](http://li.nk)"
```

Comment:
Are we testing for newlines in the URL at all? If we provide newlines and they are not removed, that will mess up the link generator because it splits over newlines and does a recursive call for each line, so it's important that we test this case despite it being a ridiculous input.

Reply:
There is now a test for newlines in links